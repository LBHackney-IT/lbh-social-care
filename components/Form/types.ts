/* eslint-disable @typescript-eslint/no-explicit-any */

import { InputHTMLAttributes } from 'react';
import { RegisterOptions } from 'react-hook-form';
import TextInput from './TextInput/TextInput';

type LabelSize = 's' | 'm' | 'l' | 'xl';
type Width = 30 | 20 | 10 | 5 | 4 | 3 | 2;
type GovGrid =
  | 'full'
  | 'half'
  | 'one-half'
  | 'one-third'
  | 'two-thirds'
  | 'one-quarter'
  | 'three-quarters';

export type Option = string | { text: string; value: string };

type FuncOption = (arg0: Record<string, string>) => Option[] | null;

export type Options = FuncOption | Option[];

export interface GenericField {
  name: string;
  label?: string;
  rules?: RegisterOptions;
  register?: (arg0: RegisterOptions) => any;
  error?: { message?: string };
  required?: boolean | string;
  hint?: string;
  labelSize?: LabelSize;
  conditionalRender?: (arg0: Record<string, any>) => boolean;
  isMulti?: boolean;
  isMultiInit?: boolean;
  isMultiTrigger?: string;
}

export interface Checkbox extends GenericField {
  options?: Option[];
}

interface StepCheckbox extends Omit<Checkbox, 'options'> {
  component: 'Checkbox';
  options?: Options;
}

type TextInputType = GenericField & InputHTMLAttributes<HTMLInputElement>;

export interface TextInput extends TextInputType {
  inputClassName?: string;
  width?: Width;
}

interface StepTextInput extends TextInput {
  component: 'TextInput';
}

export type TextInputNoType = Omit<TextInput, 'type'>;

interface StepTextInputNoType extends TextInputNoType {
  component: 'NumberInput' | 'DatePicker' | 'PhoneInput' | 'EmailInput';
}

type TextAreaType = GenericField & InputHTMLAttributes<HTMLTextAreaElement>;

export interface TextArea extends TextAreaType {
  width?: Width;
  rows?: number;
}

interface StepTextArea extends TextArea {
  component: 'TextArea';
}

type SelectType = GenericField &
  Omit<InputHTMLAttributes<HTMLSelectElement>, 'onChange'>;

export interface Select extends SelectType {
  options: Option[];
  placeHolder?: string;
  ignoreValue?: boolean;
  isUnselectable?: boolean;
  width?: Width;
  govGrid?: GovGrid;
  onChange?: (arg0: string) => string | void;
}

interface StepSelect extends Omit<Select, 'options'> {
  component: 'Select';
  options: Options;
}

type RadiosType = GenericField & InputHTMLAttributes<HTMLInputElement>;

export interface Radios extends RadiosType {
  options?: Option[];
  isRadiosInline?: boolean;
  children?: React.ReactNode;
}

interface StepRadios extends Omit<Radios, 'options'> {
  component: 'Radios';
  options?: Options;
}

export interface ObjectInput extends GenericField {
  isMulti?: boolean;
  isInline?: boolean;
  summaryInline?: boolean;
  components: Array<formBasicComponentStep>;
}

interface StepObjectInput extends ObjectInput {
  component: 'ObjectInput';
}

export interface AddressLookup extends GenericField {
  supportManualEntry?: boolean;
}

interface StepAddressLookup extends AddressLookup {
  component: 'AddressLookup';
}

export interface DateInput extends GenericField {
  format?: 'US' | 'EU';
}

interface StepDateInput extends DateInput {
  component: 'DateInput';
}

type formBasicComponentStep =
  | StepTextInput
  | StepTextInputNoType
  | StepTextArea
  | StepCheckbox
  | StepSelect
  | StepRadios
  | StepAddressLookup
  | StepDateInput;

export type FormComponentStep =
  | formBasicComponentStep
  | StepObjectInput
  | JSX.Element;

export interface FormStep {
  id: string;
  title: string;
  components: Array<FormComponentStep>;
  isMulti?: boolean;
  conditionalRender?: (arg0: Record<string, unknown>) => boolean;
}

export interface FormConfig {
  title: string;
  path: string;
  successMessage?: string;
  steps: Array<FormStep>;
}
