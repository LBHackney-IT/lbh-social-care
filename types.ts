import { AxiosError } from 'axios';

export type ErrorAPI = AxiosError;

export interface Address {
  address: string;
  postcode: string;
  uprn: string;
}

export type AgeContext = 'A' | 'B' | 'C' | undefined;

export interface Allocation {
  id: number;
  caseStatus: 'Closed' | 'Open';
  allocatedWorkerTeam: string;
  allocatedWorker: string;
  allocationStartDate: string;
  allocationEndDate?: string;
  workerType: string;
  personId: number;
  personName: string;
  personDateOfBirth: string;
  personAddress: string;
}

export interface AllocationData {
  allocations: Allocation[];
}

interface CaseFormDataBase {
  mosaic_id: number;
  first_name: string;
  last_name: string;
  worker_email: string;
  form_name_overall: string;
  form_name: string;
  context_flag: AgeContext;
  date_of_event: string | Date;
  timestamp: string | Date;
  is_historical?: true | false;
  note?: string;
  date_of_birth?: string;
  form_url?: string;
  case_note_title?: string;
  case_note_description?: string;
}

export interface AllocationCaseFormData extends CaseFormDataBase {
  form_name_overall: 'API_Allocation';
  allocation_id: number;
  created_by: string;
}

export interface DeallocationCaseFormData extends CaseFormDataBase {
  form_name_overall: 'API_Deallocation';
  allocation_id: number;
  deallocation_reason: string;
  created_by: string;
}

export interface WarningNoteCaseFormData extends CaseFormDataBase {
  form_name_overall: 'API_WarningNote';
  warning_note_id: number;
}

export interface HistoricCaseData {
  title: string;
  formName?: string;
  content?: string;
  officerName?: string;
  officerEmail?: string;
  dateOfEvent: string;
}

export interface HistoricVisitData {
  visitId: number;
  personId: number;
  visitType: string;
  plannedDateTime: string;
  actualDateTime: string;
  createdByName: string;
  createdByEmail: string;
  reasonNotPlanned?: string;
  reasonVisitNotMade?: string;
  seenAloneFlag: boolean;
  completedFlag: boolean;
}

export type CaseFormData =
  | CaseFormDataBase
  | AllocationCaseFormData
  | DeallocationCaseFormData
  | WarningNoteCaseFormData;

export interface Case {
  recordId: string;
  formName: string;
  personId: number;
  firstName?: string;
  lastName?: string;
  dateOfBirth?: string;
  officerEmail?: string;
  caseFormUrl?: string;
  caseFormTimestamp?: string;
  dateOfEvent?: string;
  caseFormData: CaseFormData;
}

export interface CaseData {
  cases: Case[] | [];
  nextCursor?: number;
}

export interface LegacyResident {
  mosaicId: number;
  firstName: string;
  lastName: string;
  uprn?: string;
  dateOfBirth: string;
  ageContext: AgeContext;
  gender: string;
  nationality?: string;
  nhsNumber: string;
  restricted?: 'Y' | 'N';
  address?: {
    address: string | null;
    postcode: string | null;
    uprn?: string;
  };
  phoneNumber?: Array<{
    phoneNumber: string;
    phoneType: string;
  }>;
}

export interface ResidentsAPI {
  residents: LegacyResident[] | [];
  nextCursor?: string;
}

/**
 * This should replace LegacyResident
 * When /residents is going to be aligned with /residents/:id
 */
export interface Resident {
  id: number;
  firstName: string;
  lastName: string;
  gender: string;
  contextFlag: AgeContext;
  createdBy: string;
  otherNames: Array<{
    firstName: string;
    lastName: string;
  }>;
  phoneNumbers: Array<{
    number: string;
    type: string;
  }>;
  title?: string;
  dateOfBirth?: string;
  ethnicity?: string;
  firstLanguage?: string;
  religion?: string;
  sexualOrientation?: string;
  nhsNumber?: number;
  emailAddress?: string;
  preferredMethodOfContact?: string;
  restricted?: 'Y' | 'N';
  dateOfDeath?: string;
  address?: {
    address: string;
    postcode: string;
    uprn?: string;
  };
}

export interface User {
  name: string;
  email: string;
  permissionFlag: AgeContext;
  hasDevPermissions: boolean;
  hasAdminPermissions: boolean;
  hasAdultPermissions: boolean;
  hasChildrenPermissions: boolean;
  hasAllocationsPermissions?: boolean;
  hasUnrestrictedPermissions?: boolean;
  isAuthorised: boolean;
}

export interface Team {
  id: number;
  name: string;
}

export interface Worker {
  id: number;
  email: string;
  firstName: string;
  lastName: string;
  role: string;
  allocationCount: number;
  teams: Team[];
}

export interface RelationshipPerson {
  id: number;
  firstName: string;
  lastName: string;
}

export interface Relationship {
  parents: Array<RelationshipPerson>;
  siblings: Array<RelationshipPerson>;
  children: Array<RelationshipPerson>;
  other: Array<RelationshipPerson>;
}

export interface RelationshipData {
  personId: number;
  personalRelationships: Relationship;
}

interface BaseNote {
  id: number;
  noteType: string;
  createdBy: string;
  startDate: string;
  reviewDate: string;
  endDate?: string;
  endedBy?: string;
  reviewedDate?: string;
  reviewedBy?: string;
  nextReviewDate?: string;
  notes: string;
  managerName: string;
  discussedWithManagerDate: string;
  status: 'closed' | 'open';
  warningNoteReviews: Array<ReviewedNote>;
}

interface DisclosedNote extends BaseNote {
  disclosedWithIndividual: true;
  disclosedDetails: string;
  disclosedDate: string;
  disclosedHow: 'Verbal' | 'Written' | 'Verbal / Written';
}

interface UndisclosedNote extends BaseNote {
  disclosedWithIndividual: false;
  undisclosedDetails: string;
}

interface ReviewedNote {
  id: number;
  warningNoteId: number;
  reviewDate: string;
  disclosedWithIndividual: boolean;
  reviewNotes: string;
  managerName: string;
  discussedWithManagerDate: string;
  createdAt: string;
  createdBy: string;
  lastModifiedAt?: string;
  lastModifiedBy: string;
}

export type WarningNote = DisclosedNote | UndisclosedNote;

export interface WarningNotes {
  notes: Array<WarningNote>;
}
