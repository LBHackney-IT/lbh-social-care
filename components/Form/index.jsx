import AddressLookup from './AddressLookup/AddressLookup';
import Autocomplete from './Autocomplete/Autocomplete';
import Checkbox from './Checkbox/Checkbox';
import DateInput from './DateInput/DateInput';
import DatePicker from './DatePicker/DatePicker';
import Radios from './Radios/Radios';
import Select from './Select/Select';
import TextInput from './TextInput/TextInput';
import TextArea from './TextArea/TextArea';
import EmailInput from './EmailInput/EmailInput';
import NumberInput from './NumberInput/NumberInput';
import PhoneInput from './PhoneInput/PhoneInput';
import ObjectInput from './ObjectInput/ObjectInput';

export {
  AddressLookup,
  Autocomplete,
  Checkbox,
  DateInput,
  DatePicker,
  Radios,
  Select,
  TextInput,
  TextArea,
  EmailInput,
  NumberInput,
  PhoneInput,
  ObjectInput,
};
