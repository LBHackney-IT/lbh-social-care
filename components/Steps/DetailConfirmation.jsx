import PropTypes from 'prop-types';
import { useRouter } from 'next/router';
import Button from 'components/Button/Button';

const DetailConfirmation = ({ formSteps, successMessage }) => {
  const router = useRouter();
  const { id } = router.query;
  if (!formSteps) return null;
  return (
    <>
      <h1 className="govuk-fieldset__legend--l gov-weight-lighter govuk-!-margin-bottom-7">
        {successMessage || 'Submission complete'}
      </h1>
      <div className="govuk-panel govuk-panel--confirmation govuk-!-margin-bottom-9">
        <h1 className="govuk-fieldset__legend--l gov-weight-lighter">
          {formSteps[0].title} details have been added
        </h1>
      </div>
      <div style={{ display: 'flex', justifyContent: 'space-between' }}>
        <Button label="Return to search" isSecondary wideButton route={`/`} />
        <Button label="View person" wideButton route={`/people/${id}`} />
      </div>
    </>
  );
};

DetailConfirmation.propTypes = {
  formData: PropTypes.shape({}).isRequired,
  formSteps: PropTypes.arrayOf(PropTypes.shape({ title: PropTypes.string }))
    .isRequired,
  formPath: PropTypes.string.isRequired,
  successMessage: PropTypes.string,
  isSummaryCollapsable: PropTypes.bool,
};

export default DetailConfirmation;
