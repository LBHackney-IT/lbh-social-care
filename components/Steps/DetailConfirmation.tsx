import { useRouter } from 'next/router';

import Button from 'components/Button/Button';
import { FormStep } from 'components/Form/types';

interface Props {
  formSteps: FormStep[];
  successMessage?: string;
}

const DetailConfirmation = ({
  formSteps,
  successMessage,
}: Props): React.ReactElement => {
  const router = useRouter();
  const { id } = router.query;
  return (
    <>
      <h1 className="govuk-fieldset__legend--l gov-weight-lighter govuk-!-margin-bottom-7">
        {successMessage || 'Submission complete'}
      </h1>
      <div className="govuk-panel govuk-panel--confirmation govuk-!-margin-bottom-9">
        <h1 className="govuk-fieldset__legend--l gov-weight-lighter">
          {formSteps[0].title} details have been added
        </h1>
      </div>
      <div style={{ display: 'flex', justifyContent: 'space-between' }}>
        <Button label="Return to search" isSecondary wideButton route={`/`} />
        <Button label="View person" wideButton route={`/people/${id}`} />
      </div>
    </>
  );
};

export default DetailConfirmation;
