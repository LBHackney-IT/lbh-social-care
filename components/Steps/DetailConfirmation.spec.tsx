import { render } from '@testing-library/react';

import DetailConfirmation from './DetailConfirmation';
import { FormStep } from 'components/Form/types';

jest.mock('next/router', () => ({
  useRouter: () => ({
    query: { id: 'foo123456bar' },
  }),
}));

describe('Detail Confirmation component', () => {
  const props = {
    formData: {
      bar_input: 'foo',
    },
    formSteps: [
      {
        components: [
          {
            component: 'TextInput',
            name: 'bar_input',
            width: 30,
            label: 'Foo',
          },
        ],
        id: 'foo-bar',
        title: 'Foo Bar',
      },
    ] as FormStep[],
    successMessage: 'Done!',
    formPath: 'foo/bar-foo',
  };

  it('should render properly', () => {
    const { asFragment, getByText } = render(<DetailConfirmation {...props} />);
    expect(asFragment()).toMatchSnapshot();
    expect(getByText(props.successMessage)).toBeInTheDocument();
  });
});
