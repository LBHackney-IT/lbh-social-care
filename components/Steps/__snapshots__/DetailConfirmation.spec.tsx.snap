// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Detail Confirmation component should render properly 1`] = `
<DocumentFragment>
  <h1
    class="govuk-fieldset__legend--l gov-weight-lighter govuk-!-margin-bottom-7"
  >
    Done!
  </h1>
  <div
    class="govuk-panel govuk-panel--confirmation govuk-!-margin-bottom-9"
  >
    <h1
      class="govuk-fieldset__legend--l gov-weight-lighter"
    >
      Foo Bar details have been added
    </h1>
  </div>
  <div
    style="display: flex; justify-content: space-between;"
  >
    <button
      class="govuk-button lbh-button govuk-button--secondary lbh-button--secondary"
      data-module="govuk-button"
      style="min-width: 200px;"
    >
      Return to search
    </button>
    <button
      class="govuk-button lbh-button"
      data-module="govuk-button"
      style="min-width: 200px;"
    >
      View person
    </button>
  </div>
</DocumentFragment>
`;
